### PR TITLE
Add rank loss operator

### LINKED ISSUE #4065: Rank Loss Operator

### PR DESCRIPTION
**Pull request by kuke**
Resolve #4065

### CODE REVIEW

File: python/paddle/v2/framework/tests/test_rank_loss_op.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from op_test import OpTest
+
+
+class TestReshapeOp(OpTest):
+    def setUp(self):
+        self.op_type = "rank_loss"
+        num = 5
+        # P = {0, 1.0} or {0, 0.5, 1.0}
+        P = np.random.randint(0, 2, size=(num, num)).astype("float32")
+        Oi = np.random.random((num, num)).astype("float32")
```
Comment (Yancey1989) [inline python/paddle/v2/framework/tests/test_rank_loss_op.py]: Local variable  name should be `lower_with_under`, https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming

File: paddle/operators/rank_loss_op.cc [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+\f[
+  C_{i,j} = -\tilde{P_{ij}} * o_{i,j} + log(1 + e^{o_{i,j}}) \\
+  o_{i,j} =  o_i - o_j  \\
+  \tilde{P_{i,j}} = \left \{0, 0.5, 1 \right \} \ or \ \left \{0, 1 \right \}
+\f]
+
+A detailed explanation about these notations can be found in
+
+[1]. Chris Burges, Tal Shaked, Erin Renshaw, et al. Learning to
```
Comment (Yancey1989) [inline paddle/operators/rank_loss_op.cc]: Maybe we can add the link here.

File: paddle/operators/rank_loss_op.cc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("P"), "Input(P) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oi"), "Input(Oi) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oj"), "Input(Oj) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) shouldn't be null.");
+    auto dims = ctx.Input<framework::Tensor>("P")->dims();
+    ctx.Output<framework::Tensor>(framework::GradVarName("P"))->Resize(dims);
+    ctx.Output<framework::Tensor>(framework::GradVarName("Oi"))->Resize(dims);
+    ctx.Output<framework::Tensor>(framework::GradVarName("Oj"))->Resize(dims);
```
Comment (Canpio) [inline paddle/operators/rank_loss_op.cc]: Gradient Op's output(the gradient of forward op's inputs) can be `nullptr`, which means they are not necessary for backward. So we shall assert it is not `nullptr` before `Resize`.

See https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/cos_sim_op.cc#L146

File: paddle/operators/rank_loss_op.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    auto* d_oi = ctx.Output<framework::Tensor>(framework::GradVarName("Oi"));
+    auto* d_oj = ctx.Output<framework::Tensor>(framework::GradVarName("Oj"));
+    auto* d_p = ctx.Output<framework::Tensor>(framework::GradVarName("P"));
+
+    auto* d_out = ctx.Input<framework::Tensor>(framework::GradVarName("Out"));
+    auto* p_t = ctx.Input<framework::Tensor>("P");
+    auto* oi_t = ctx.Input<framework::Tensor>("Oi");
+    auto* oj_t = ctx.Input<framework::Tensor>("Oj");
+
+    d_oi->mutable_data<T>(ctx.GetPlace());
```
Comment (Canpio) [inline paddle/operators/rank_loss_op.h]: Outputs of gradient Op may be `nullptr`. If so, it means that they are useless for backward and we don't need to compute them. 

see https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/cos_sim_op.h#L104 for an example.

File: python/paddle/v2/framework/tests/test_rank_loss_op.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import unittest
+import numpy as np
+from op_test import OpTest
+
+
+class TestReshapeOp(OpTest):
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_rank_loss_op.py]: Why the class name is `TestReshapeOp`?

File: python/paddle/v2/framework/tests/test_rank_loss_op.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        Oj = np.random.random((num, num)).astype("float32")
+        O = Oi - Oj
+        Out = np.log(1.0 + np.exp(O)) - P * O
+        self.inputs = {'P': P, 'Oi': Oi, 'Oj': Oj}
+        self.outputs = {'Out': Out}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_rank_loss_op.py]: Add some `check_grad_ignore_XXX` tests if posiible.
In `check_grad_ignore_XXX` tests, ignored variables' gradients will be set `nullptr` and your kernel should not compute it. 

Example: https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/framework/tests/test_mul_op.py#L21

File: paddle/operators/rank_loss_op.cc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+class RankLossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  RankLossOpMaker(framework::OpProto *proto,
+                  framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("P", "The desired target values for posteriors.");
+    AddInput("Oi", "The model output for item i.");
+    AddInput("Oj", "The model output for item j.");
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/rank_loss_op.cc]: Done

File: paddle/operators/rank_loss_op.cc (hunk 2 above)
Comment (kuke) [reply paddle/operators/rank_loss_op.cc]: Done

File: paddle/operators/rank_loss_op.cc (hunk 3 above)
Comment (kuke) [reply paddle/operators/rank_loss_op.cc]: Done

File: paddle/operators/rank_loss_op.h (hunk 4 above)
Comment (kuke) [reply paddle/operators/rank_loss_op.h]: Done

File: python/paddle/v2/framework/tests/test_rank_loss_op.py (hunk 5 above)
Comment (kuke) [reply python/paddle/v2/framework/tests/test_rank_loss_op.py]: A typo. Fixed

File: python/paddle/v2/framework/tests/test_rank_loss_op.py (hunk 1 above)
Comment (kuke) [reply python/paddle/v2/framework/tests/test_rank_loss_op.py]: Done

File: python/paddle/v2/framework/tests/test_rank_loss_op.py (hunk 6 above)
Comment (kuke) [reply python/paddle/v2/framework/tests/test_rank_loss_op.py]: Done

**Pull request merged**